### PR TITLE
Drop execution_date unique constraint on DagRun

### PR DESCRIPTION
**Pull request by uranusjr**
The column has also been renamed to logical_date, although the Python model is not changed. This allows us to not need to fix all the Python code at once (we'll do that later), but still do the two changes in one migration instead of two.

Note that this one is not breaking—the metadatabase interface is not user-facing, we’re only removing a check that stopped impossible things.

### CODE REVIEW

File: airflow/utils/db.py [hunk 1]
```diff
@@ -1194,7 +1194,7 @@ def check_run_id_null(session: Session) -> Iterable[str]:
     invalid_dagrun_filter = or_(
         dagrun_table.c.dag_id.is_(None),
         dagrun_table.c.run_id.is_(None),
-        dagrun_table.c.execution_date.is_(None),
+        dagrun_table.c.logical_date.is_(None),
```
Comment (uranusjr) [inline airflow/utils/db.py]: This is for data integrity checks in `upgradedb`. But I’m thinking maybe we can just drop the checks entirely?

If we _require_ users to upgrade to 2.10 (or even 2.11) before upgrading, they should have already run those checks and have data integrity (enforced by the database) when they upgrade to 3.0.

(Same for the column renames below.)

Comment (uranusjr) [reply airflow/utils/db.py]: I raised #41850 to remove the checks entirely.

**Review by sunank200 — APPROVED** [review]

**Review by sunank200 — COMMENTED** [review]
I think the tests has to change as well

**Comment by uranusjr** [comment]
Fixed. If this passes, I think it can be merged first.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 2]
```diff
@@ -172,7 +172,7 @@ def _fetch_dag_runs(
         query = query.where(DagRun.updated_at <= updated_at_lte)
 
     total_entries = get_query_count(query, session=session)
-    to_replace = {"dag_run_id": "run_id"}
+    to_replace = {"dag_run_id": "run_id", "execution_date": "logical_date"}
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py:175]: This workaround is due to `apply_sorting` using the _raw SQL column name_, not the SQLAlchemy definition, so we need to explicitly fix the name separately. Eventually we should rename the argument on this API to `logical_date` instead.

File: airflow/www/utils.py [hunk 3]
```diff
@@ -521,7 +521,7 @@ def dag_run_link(attr):
 
 
 def _get_run_ordering_expr(name: str) -> ColumnOperators:
-    expr = DagRun.__table__.columns[name]
+    expr = DagRun.__mapper__.columns[name]
```
Comment (uranusjr) [inline airflow/www/utils.py:525]: `__table__` uses the real column names, while `__mapper__` gives the names on the model class.

**Review by kaxil — APPROVED** [review]
It isn’t breaking change but might be worth mentioning in Towncrier notes as a note-worthy change if someone is using combination of dag_id & execution_date to retrieve DagRuns

**Comment by uranusjr** [comment]
I think it is more appropriate to attach the news fragment on the next PR that will rename the attribute on DagRun instead. That one is the actual breaking change most people will notice. This PR only affects the SQL layer, not Python.

**Pull request merged**